identity/src/public_key.rs

```diff
@@ -13,63 +13,122 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-
-use crate::public_key::PKError::{FromBase58Error, FromBytesError};
+use crate::ed25519;
+use crate::rsa;
+use crate::secp256k1;
+use crate::error::*;
 use crate::signature::Signature;
 
-use core::fmt::Debug;
-use ed25519_dalek::SignatureError;
 use serde::{Deserialize, Serialize};
-use thiserror::Error as ThisError;
-use libp2p_core::identity;
-
-#[derive(ThisError, Debug)]
-pub enum PKError {
-    #[error("Cannot decode public key from bytes: {0}")]
-    FromBytesError(#[source] SignatureError),
-    #[error("Cannot decode public key from base58 format: {0}")]
-    FromBase58Error(#[source] bs58::decode::Error),
+
+/// The public key of a node's identity keypair.
+#[repr(u8)]
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub enum PublicKey {
+    /// A public Ed25519 key.
+    Ed25519(ed25519::PublicKey),
+    #[cfg(not(target_arch = "wasm32"))]
+    /// A public RSA key.
+    Rsa(rsa::PublicKey),
+    /// A public Secp256k1 key.
+    Secp256k1(secp256k1::PublicKey)
 }
 
-#[derive(Copy, Clone, Default, Eq, PartialEq, Serialize, Deserialize)]
-pub struct PublicKey(pub(crate) ed25519_dalek::PublicKey);
+impl PublicKey {
+    /// Verify a signature for a message using this public key, i.e. check
+    /// that the signature has been produced by the corresponding
+    /// private key (authenticity), and that the message has not been
+    /// tampered with (integrity).
+    pub fn verify(&self, msg: &[u8], sig: &Signature) -> bool {
+        use PublicKey::*;
+        match self {
+            Ed25519(pk) => pk.verify(msg, sig.to_bytes().as_slice()),
+            #[cfg(not(target_arch = "wasm32"))]
+            Rsa(pk) => pk.verify(msg, sig.to_bytes().as_slice()),
+            Secp256k1(pk) => pk.verify(msg, sig.to_bytes().as_slice())
+        }
+    }
 
-impl Debug for PublicKey {
-    fn fmt(&self, f: &mut ::core::fmt::Formatter<'_>) -> ::core::fmt::Result {
-        write!(f, "{:?}", self.0)
+    pub fn encode(&self) -> Vec<u8> {
```

Comment:
I'd return `Cow<[u8]>` here to avoid copying when possible (for Ed25519 and Secp256k1). Something along these lines:
```suggestion
    pub fn encode(&self) -> Cow<[u8], '_> {
```